### PR TITLE
Reviews, comments and discussion

### PR DESCRIPTION
**Pull request by dougiesquire**
I've set up this pull request to an empty branch so that we can use github's handy review tools to review all of the code in BigData. I'll give an example performing a review below.

To open this PR with ReviewNB (for reviewing .ipynb files), go to: https://app.reviewnb.com/ACDguide/BigData/pull/10/

If others have suggestions for how to do this better, please chime in.

### CODE REVIEW

File: BigData/data_storage.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+## NetCDF-Zarr
+
+Since [netCDF v4.8.0](https://www.unidata.ucar.edu/blogs/developer/entry/overview-of-zarr-support-in), support for zarr as a storage backend has been added.
+
+* When netCDF is built with zarr enabled, it is possible to write a netCDF file out in Zarr format. 
+* At this stage, reading of Zarr files is not yet enabled by the python NetCDF library, but that will be added in due course.
+* netCDF 4.8.0 does not (yet) support for zip output, so at this stage, writing a netCDF file out in zarr format creates many file objects (one per chunk)
+* "nczarr" does not support unlimited dimensions like HDF5 does, as such netCDFs must have all dimensions of defined length. 
+
+### What we found
```
Comment (dougiesquire) [inline BigData/data_storage.md:63]: I'm really interested in your testing and I think it's really valuable to have this documented, but I wonder if this section is a little too specific for the guide. I wonder if we could instead include a very brief summary of our (your) experiences, or even just a statement that nczarr is still very new and lacking some key functionality (which you've kind of already said above).

File: BigData/data_storage.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+`ncgen: NetCDF: Attempt to use feature that was not turned on when netCDF was built.
+        (../../ncgen/genbin.c:genbin_netcdf:63)`
+
+Findings: 
+* It is not viable to translate existing netCDF to zarr-backed netCDF, because we liberally use unlimited time dimensions.
+* It is not a good idea to write much data to zarr format using this tool at the moment because it doesn't support the zip archiving, so uses a lot of inodes, which will affect quotas.
+* It is important to watch this library as it (rapidly) evolves to better supoort zarr read/write.
+
+NetCDF remains the standards-compliant file format for climate data which is broadly supported by many tools, but lack optimal parallel performance. As this library matures it should be possible eventually to store our data in standards- and tools-compliant netCDF format but with zarr performance.
+
```
Comment (dougiesquire) [inline BigData/data_storage.md:103]: So far, we're very NetCDF- and zarr-centric. These are the two key formats that I interact with, but I'm sure there're others that we should include. Where can we get that broader perspective>

**Review by dougiesquire — COMMENTED** [review]
@hot007, some thoughts/comments. 

I'm providing my comments as a review to see if that works for people. I think it could be an okay way to keep comments/discussion tied to specific parts of the document, but let me know how it goes for you. Also happy to just edit the document directly and put discussion into issues.

File: BigData/computations.ipynb [hunk 3]
```diff
@@ -0,0 +1,750 @@
+{
```
Comment (dougiesquire) [inline BigData/computations.ipynb:1]: <p>Could we include a description of what the notebook contains, e.g.</p><p>"This notebook introduces a number of commonly-used tools for performing computations with large climate datasets, and demonstrates how common tasks are carried out on chunked data using tools like <code>dask</code>"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ACDguide/BigData/pull/10/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (dougiesquire) [inline BigData/computations.ipynb:1]: <p>I wonder if having this code block right at the start with no context might be a little intimidating/confusing to some readers? Could we add a header to this code block to make it clear that users don't need to worry about this, e.g.</p><h3>Set up a function to help us draw some diagrams below</h3>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ACDguide/BigData/pull/10/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (dougiesquire) [inline BigData/computations.ipynb:1]: <p>Missing space before </p><h3>Command-line tools </h3>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ACDguide/BigData/pull/10/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (dougiesquire) [inline BigData/computations.ipynb:1]: <p>Missing space before</p><h3>Percentiles / Median </h3><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ACDguide/BigData/pull/10/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (dougiesquire) [inline BigData/computations.ipynb:1]: <p>Resampling could also include increasing the frequency. E.g. xarray resample can handle upsampling</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/ACDguide/BigData/pull/10/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

**Comment by hot007** [comment]
So how do you want to do this @dougiesquire , if I want to make changes you suggest do I go and edit the original file and re-commit, then accept your comment here? I had a look at reviewNB and honestly it was messy and confusing with so many reviews there, I actually couldn't figure out where these comments had gone!
If I don't use reviewNB, is there a way to make your suggested edits in this PR or should i go and do it on main and then mark as done here?

File: BigData/data_storage.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+Traditionally, climate data have been stored as GRIB (GRIdded Binary) or NetCDF (Network Common Data Format) files. These file formats support large array data and metadata. NetCDF files can be subset to extract data over specified dimension ranges, including remotely over the internet using the [OPeNDAP protocol](https://www.opendap.org/) via THREDDS, Hyrax or PyDAP. NetCDF4 supports data "chunking" to optimise performance for particular read patterns.
+
+With the rise of cloud computing, the need for data formats that can perform more optimally with parallelised access patterns has emerged. Zarr (zipped archive) is an alternate way to store climate data, whereby each netCDF "chunk" is written to a unique "object", thereby permitting much higher levels of efficient parallelised access. However, on a traditional filesystem, each chunk is written as an individual file, which can affect quota limits. To avoid this problem, the Zarr representation of a dataset may be stored in a zip file so they are reduced to a single file-object, without loss of performance with tools like `xarray`.
```
[the remark this answers is not in the document]
Comment (hot007) [reply BigData/data_storage.md:12]: Hmm, true. I'd like to mention cloud because that's really what triggered it (similar to COG which also performs better than not-CoG even on not-cloud I understand) but you're right, I could rephrase this to emphasise that the need was there anyway.
Also thanks re zipping/performance, will modify accordingly.

File: BigData/data_storage.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+An interesting recent development is support for zarr as a storage back-end for netCDF, thereby potentially offering the best of both worlds.
+
+### NetCDF
+[NetCDF](https://www.unidata.ucar.edu/software/netcdf/) is the data format most commonly used in climate science. It is an open data format with full self-described metadata, though it is up to the data creator to ensure the metadata meets any relevant standards, such as the [CF Convention](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html).
+
+### Zarr
+[Zarr](https://zarr.readthedocs.io/en/stable/) is a new data format that is optimised for cloud interaction with data. Zarr is a portmanteau of "zipped archive", it is used by python's `xarray` to more performantly store data, by writing a separate file object for each data chunk. This makes data indexing and access much faster than on a monolithic netCDF file, and performance improvements are seen on local HPC storage as well as cloud.
+
+### NetCDF: what's under the hood?
```
[the remark this answers is not in the document]
Comment (hot007) [reply BigData/data_storage.md:23]: Yes. I had other plans but then couldn't find the will to write an essay about how netCDF works :p

File: BigData/data_storage.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+## Advice on writing large datasets efficiently
+
+CLEX CMS team have produced [some useful advice](http://climate-cms.wikis.unsw.edu.au/NetCDF_Compression_Tools) about storage of netCDF data. That page details 
+* deflation, 
+* shuffling, 
+* chunking, and 
+* command line tools to control file storage structure (some are specific to [NCI](https://nci.org.au/)).
+
+All of these things can have massive impacts on file performance for both regular access but in particular parallelised access using `dask`. Data will only be performant with a tool like `dask` if it is structured appropriately for the read patterns, and chunk arguments supplied to `xarray` must align with the chunk sizes the data is physically stored in, otherwise you can end up with *worse* performance.
```
[the remark this answers is not in the document]
Comment (hot007) [reply BigData/data_storage.md:48]: Yes, it's not particularly intuitive that the two ways to use dask are to explicitly use dask, or to use xarray with chunks specified, and more careful phrasing is called for. Thanks.

File: BigData/data_storage.md (hunk 1 above)
Comment (hot007) [reply BigData/data_storage.md:63]: Yeah, I wonder if we could make it a sub-page - I fully agree, it is of approximately zero relevance to someone looking for reference information, however for our immediate community there's logic to having a place where usability and performance of tools and libraries are documented. What would you say to keeping it but not on this page, rather burying it a bit (assuming jupyterbooks work that way?)?

File: BigData/data_storage.md (hunk 2 above)
Comment (hot007) [reply BigData/data_storage.md:103]: That's part of why I included the table of file formats. I do wonder if we should talk more about Grib, also whether we should include `.pp` files produced by the ACCESS model (and others), but part of that is determining who our audience is? 
If we're covering all of Earth Science, then we certainly need to go much broader into GeoTIFF, point cloud formats, other vector data formats, as well as myriad medium-scale raster formats (CSV, .mat, etc). 
If our focus is purely on climate data then actually I think a focus on netCDF, with discussion of zarr, HDF? and grib? is appropriate, as community standards/tools and conventions mean that it really is the format that is most used and people most need to understand. At least that's my (limited/biased?) view.

File: BigData/data_storage.md (hunk 1 above)
Comment (paigem) [reply BigData/data_storage.md:63]: I think it makes sense to keep this section for now, but agree with @hot007 that maybe we can make it be a sub-page. We could even title the sub-page something like "In-depth nczarr testing" or something that implies that the sub-page is very specific. Or even something along the lines of "Appendix: nczarr testing" (not sure if appendix is the best word choice here, but I think it conveys the idea that the section is only for those interesting diving deeper into the topic).

I guess my philosophy with this guide right now is to include anything we think might be useful, and later on in the process we can decide to remove/edit sections if we want a clean, polished product. But perhaps that is just creating more work for us later on?

File: BigData/data_storage.md (hunk 2 above)
Comment (paigem) [reply BigData/data_storage.md:103]: I agree with @hot007 that for now I think we can just focus on climate data (after all this is a "Climate Data Guide"). But, perhaps we could list other file formats, maybe with a few words describing what they are used for? For instance, I don't even know what Grib or .pp files are, and knowing that they exist is probably useful for (at the bare minimum) communicating with earth scientists.

File: BigData/computations.ipynb (hunk 3 above)
Comment (paigem) [reply BigData/computations.ipynb:1]: <p>To go even further, is there a way to toggle the viewing of this code block? E.g. using html/markdown commands like <code>&lt;details&gt;&lt;/details&gt;</code> that show the code block if a user clicks on the arrow?</p>

**Comment by paigem** [comment]
@hot007 I'm not sure, but I think the comments that @dougiesquire made on your notebook were before I added ReviewNB as a feature of this repo. So that may be why you couldn't find your comments in Review NB. (But I'm very new to ReviewNB so I may be missing the mark here.)

Also I just saw that @dougiesquire said to ignore the comments in this thread - whoops!

Moving forward, if we want to make comments on the content of the repo, what do you suggest is the best way to do this @dougiesquire? I like the idea of using Github's review tools, but am so new to them that I just don't know how to best make use of them.

**Comment by dougiesquire** [comment]
Thanks @hot007 and @paigem!

To be clear, this is not a standard way of using PRs, but I thought it might be a handy way to provide people access to github's review features so that we can make comments and have discussions attached to specific lines of code/documentation.

**What is typically done:** a user submits a PR to merge their branch with the `main` branch. Their branch will be different in some way than the `main` branch (e.g. include a new bug fix). Within the "Files changed" tab of the PR (see above), github will flag new/removed lines of code relative to the `main` branch for other developers to review and comment on. In response to these reviews, the owner of the PR can push new commits to their branch until the reviews are happy for the branch to be merged with `main`. Once the branch is merged, the PR is closed.

**What I have done here:** this is a PR to merge the `main` branch into and empty branch. This means that all files within the `main` branch of the BigData repo are available for all to review under the "Files changed" tab of the PR. I propose we leave this PR open so that the latest version of the BigData repo is alway available to review. To make any changes that come from discussion in the reviews, these just need to be made to the `main` branch and they will also appear in this PR. One could do this by opening a separate PR (if you're comfortable doing so - probably unnecessary in most cases), using @ScottWales's handy `"."` trick, or by simply editing the file directly within github (since many of our files are just Markdown anyway - just click the little pencil in the top right of the file you want to edit). Regarding ReviewNB, this is only for reviewing .ipynb files (of which we have a few). These don't render nicely in the "Files changed" tab, so users can instead review them by following the link I provide above.

This may be a very clunky way to do this and I'm happy to ditch it. My preference would be for everyone to use PRs and reviews in the "usual" way, but this seemed like a way to enable those that wanted to access github review tools.

File: BigData/computations.ipynb (hunk 3 above)
Comment (ScottWales) [reply BigData/computations.ipynb:1]: That's what it's doing in the rendered output - see  https://acdguide.github.io/BigData/computations.html (it uses jupyter cell tags to do this). I'll add a comment to the cell though to say this is just for visualisation

File: BigData/data_storage.md (hunk 1 above)
Comment (hot007) [reply BigData/data_storage.md]: Sub-page created and ToC modified accordingly. I think this works.

File: BigData/data_storage.md (hunk 2 above)
Comment (hot007) [reply BigData/data_storage.md:62]: Added a few more words and links to the table of other formats for your edification @paigem , hope that is useful :)

**Comment by dougiesquire** [comment]
Closing since we've agreed to all use PRs to contribute
